File: src/reporter.ts
import chalk from 'chalk';
import { CheckResult } from './types.js';
import { RunnerResult } from './runner.js';

export interface ReporterOptions {
  verbose?: boolean;
  json?: boolean;
}

const CHECK_ICON = chalk.green('✓');
const FAIL_ICON = chalk.red('✗');
const PENDING_ICON = chalk.yellow('○');

function formatDuration(ms: number): string {
  if (ms < 1) return '<1ms';
  if (ms < 1000) return `${Math.round(ms)}ms`;
  return `${(ms / 1000).toFixed(1)}s`;
}

function padRight(str: string, len: number): string {
  return str.padEnd(len);
}

export function printHeader(): void {
  console.log(chalk.bold('\nDrip Health Check'));
  console.log(chalk.gray('═'.repeat(47)));
}

export function printCheckStart(name: string): void {
  process.stdout.write(`   ${PENDING_ICON} ${padRight(name, 22)}`);
}

export function printCheckResult(result: CheckResult, options: ReporterOptions = {}): void {
  // Clear the line and rewrite
  process.stdout.write('\r');

  const icon = result.success ? CHECK_ICON : FAIL_ICON;
  const duration = formatDuration(result.duration);
  const durationStr = result.duration > 0 ? chalk.gray(duration.padStart(8)) : chalk.gray('-'.padStart(8));

  console.log(`   ${icon} ${padRight(result.name, 22)}${durationStr}    ${result.message}`);

  if (!result.success && result.details && options.verbose) {
    console.log(chalk.gray(`     └─ ${result.details}`));
  }

  if (!result.success && result.suggestion) {
    console.log(chalk.yellow(`     └─ Suggestion: ${result.suggestion}`));
  }
}

export function printSummary(result: RunnerResult): void {
  console.log(chalk.gray('═'.repeat(47)));

  const total = result.passed + result.failed;
  const passedStr = chalk.green(`${result.passed}/${total} checks passed`);
  const totalTime = formatDuration(result.totalDuration);

  console.log(`   ${passedStr}${' '.repeat(20)}Total: ${totalTime}`);

  if (result.failed === 0) {
    console.log(chalk.green.bold(`\n   Status: HEALTHY ✓\n`));
  } else {
    console.log(chalk.red.bold(`\n   Status: UNHEALTHY ✗\n`));
  }
}

export function printJson(result: RunnerResult): void {
  const output = {
    status: result.failed === 0 ? 'healthy' : 'unhealthy',
    checks: result.results.map(r => ({
      name: r.name,
      success: r.success,
      duration_ms: Math.round(r.duration),
      message: r.message,
      ...(r.details && { details: r.details }),
      ...(r.suggestion && { suggestion: r.suggestion }),
    })),
    summary: {
      total: result.passed + result.failed,
      passed: result.passed,
      failed: result.failed,
      duration_ms: Math.round(result.totalDuration),
    },
  };

  console.log(JSON.stringify(output, null, 2));
}

export class Reporter {
  private options: ReporterOptions;

  constructor(options: ReporterOptions = {}) {
    this.options = options;
  }

  start(): void {
    if (!this.options.json) {
      printHeader();
    }
  }

  onCheckStart(name: string): void {
    if (!this.options.json) {
      printCheckStart(name);
    }
  }

  onCheckComplete(result: CheckResult): void {
    if (!this.options.json) {
      printCheckResult(result, this.options);
    }
  }

  finish(result: RunnerResult): void {
    if (this.options.json) {
      printJson(result);
    } else {
      printSummary(result);
    }
  }
}

File: python/reporter.py
"""Output formatting matching TypeScript version."""
import json
import sys
from typing import List
from .types import Check, CheckResult


class Reporter:
    """Handles check output formatting."""

    # ANSI color codes
    GREEN = "\033[32m"
    RED = "\033[31m"
    YELLOW = "\033[33m"
    RESET = "\033[0m"
    BOLD = "\033[1m"

    def __init__(self, verbose: bool = False, json_output: bool = False):
        """Initialize reporter.

        Args:
            verbose: Show detailed output
            json_output: Output as JSON instead of text
        """
        self.verbose = verbose
        self.json_output = json_output
        self.results: List[dict] = []
        # Check if terminal supports colors
        self.use_colors = sys.stdout.isatty() and not json_output

    def _color(self, text: str, color: str) -> str:
        """Apply color to text if colors are enabled."""
        if self.use_colors:
            return f"{color}{text}{self.RESET}"
        return text

    def start(self):
        """Called at the start of check execution."""
        if not self.json_output:
            print("\n🔍 Drip SDK Health Check (Python)\n")
            print("=" * 50)

    def on_check_start(self, check: Check):
        """Called when a check starts.

        Args:
            check: The check that is starting
        """
        if not self.json_output:
            print(f"\n▶ {check.name}: {check.description}")

    def on_check_complete(self, result: CheckResult):
        """Called when a check completes.

        Args:
            result: The result of the completed check
        """
        if self.json_output:
            self.results.append({
                "name": result.name,
                "success": result.success,
                "duration": round(result.duration, 2),
                "message": result.message,
                "details": result.details,
                "suggestion": result.suggestion
            })
        else:
            if result.success:
                status = self._color("✓", self.GREEN)
            else:
                status = self._color("✗", self.RED)

            duration_str = f"({result.duration:.0f}ms)"
            print(f"  {status} {result.message} {duration_str}")

            if self.verbose and result.details:
                print(f"    Details: {result.details}")

            if not result.success and result.suggestion:
                print(f"    💡 {result.suggestion}")

    def finish(self, results: List[CheckResult]):
        """Called when all checks are complete.

        Args:
            results: List of all check results
        """
        passed = sum(1 for r in results if r.success)
        failed = len(results) - passed
        total = len(results)

        if self.json_output:
            output = {
                "results": self.results,
                "summary": {
                    "total": total,
                    "passed": passed,
                    "failed": failed
                }
            }
            print(json.dumps(output, indent=2))
        else:
            print("\n" + "=" * 50)

            if failed > 0:
                status = self._color(f"📊 Results: {passed}/{total} passed ({failed} failed)", self.YELLOW)
            else:
                status = self._color(f"📊 Results: {passed}/{total} passed ✓", self.GREEN)

            print(f"\n{status}")
